common/crypto/platform/openssl/crypto.cpp

```diff
@@ -139,6 +150,96 @@ ExpectedPrivateKey PrivateKey::LoadFromPEM(const string &private_key_path) {
 	return PrivateKey::LoadFromPEM(private_key_path, "");
 }
 
+ExpectedPrivateKey PrivateKey::Generate(const unsigned int bits, const unsigned int exponent) {
+#ifdef MENDER_CRYPTO_OPENSSL_LEGACY
+	auto pkey_gen_ctx = unique_ptr<EVP_PKEY_CTX, void (*)(EVP_PKEY_CTX *)>(
+		EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, nullptr), pkey_ctx_free_func);
+
+	int ret = EVP_PKEY_keygen_init(pkey_gen_ctx.get());
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Initialization failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	ret = EVP_PKEY_CTX_set_rsa_keygen_bits(pkey_gen_ctx.get(), bits);
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Parameters setting failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	auto exponent_bn = unique_ptr<BIGNUM, void (*)(BIGNUM *)>(BN_new(), bn_free);
+	ret = BN_set_word(exponent_bn.get(), exponent);
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Parameters setting failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+
+	ret = EVP_PKEY_CTX_set_rsa_keygen_pubexp(pkey_gen_ctx.get(), exponent_bn.get());
+	if (ret != OPENSSL_SUCCESS) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to generate a private key. Parameters setting failed: "
+				+ GetOpenSSLErrorMessage()));
+	}
+	exponent_bn.release();
```

Comment:
It needs to be here, otherwise it will try to destroy an object managed by somebody else. My original gut reaction was that maybe this should be used inline instead of `.get()`, but then I realized that @lluiscampos did it the only correct way: We need to release *after* the call above, because if it fails, we still need to keep the ownership.

Reply:
Ahh, now I see. ignore me